Flaky test: TestRelayBasedSweep fails frequently

```
--- FAIL: TestRelayBasedSweep (10.03s)
	Error Trace:	idle_sweep_test.go:62
			idle_sweep_test.go:225
			test_server.go:452
			test_server.go:127
			idle_sweep_test.go:202
	Error:      	Some connections are still open: %s
	Test:       	TestRelayBasedSweep
	Messages:   	relay-40596: 2 open, testService-43024: 1 open, testService-client-242: 1 open
```

Logs:
```
	<autogenerated>:1: Debug logs:
		20:44:36.400669 [I] Created new channel. [{serviceName testService} {process testService-41529} {chID 402} {channelPtr 0xc420279380} {createdStack goroutine 3149 [running]:
		github.com/uber/tchannel-go.getStacks(0xc420279300, 0xb34f32, 0xb, 0xc4214d8dd0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:435 +0xb7
		github.com/uber/tchannel-go.registerNewChannel(0xc420279380)
			/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x6e
		github.com/uber/tchannel-go.NewChannel(0xb34f32, 0xb, 0xc420211e40, 0xb3261b, 0x1, 0xc421155eca)
			/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:292 +0xe78
		github.com/uber/tchannel-go/testutils.NewServerChannel(0xc420211ce0, 0xc4210ce690, 0xc420211ce0, 0x4159b0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:51 +0x615
		github.com/uber/tchannel-go/testutils.newServer(0xe34160, 0xc4210ce690, 0xc420211ce0, 0xc4214d9310)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:71 +0x65
		github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc4210ce690, 0xb524f8, 0xc420211ce0, 0x0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:262 +0x72
		github.com/uber/tchannel-go/testutils.(*TestServer).NewServer(0xc4210ce690, 0xc4214d9940, 0xc420498f30)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:222 +0x104
		github.com/uber/tchannel-go/testutils.NewTestServer(0xe342a0, 0xc420124690, 0xc4214d9940, 0x0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:90 +0x1d5
		github.com/uber/tchannel-go/testutils.withServer(0xe342a0, 0xc420124690, 0xc4214d9940, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:447 +0x51
		github.com/uber/tchannel-go/testutils.WithTestServer(0xe342a0, 0xc420124690, 0xc4205cccd8, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:127 +0x2b1
		github.com/uber/tchannel-go_test.TestRelayBasedSweep(0xc420124690)
			/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:202 +0x6d8
		testing.tRunner(0xc420124690, 0xb52a70)
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16d
		created by testing.(*T).Run
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:789 +0x569
		}]
		20:44:36.400818 [I] Starting idle connections poller. [{serviceName testService} {process testService-41529} {chID 402} {idleCheckInterval 30s} {maxIdleTime 3m0s}]
		20:44:36.401056 [I] Channel is listening. [{serviceName testService} {process testService-41529} {chID 402} {hostPort 127.0.0.1:41529}]
		20:44:46.411838 [D] TEST: Test function complete [{serviceName testService} {process testService-41529} {chID 402} {hostPort 127.0.0.1:41529}]
		20:44:46.418247 [D] TEST: TestServer is closing and verifying channel [{serviceName testService} {process testService-41529} {chID 402} {hostPort 127.0.0.1:41529}]
		20:44:46.418284 [I] Channel.Close called. [{serviceName testService} {process testService-41529} {chID 402} {hostPort 127.0.0.1:41529}]
		20:44:46.418425 [I] Stopping idle connections poller. [{serviceName testService} {process testService-41529} {chID 402} {hostPort 127.0.0.1:41529}]
		20:44:46.418536 [I] Channel closed. [{serviceName testService} {process testService-41529} {chID 402} {hostPort 127.0.0.1:41529}]
	<autogenerated>:1: Debug logs:
		20:44:36.401756 [I] Created new channel. [{serviceName relay} {process relay-40596} {chID 403} {channelPtr 0xc420279520} {createdStack goroutine 3149 [running]:
		github.com/uber/tchannel-go.getStacks(0xc420279500, 0xb32f09, 0x5, 0xc4214d8d60)
			/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:435 +0xb7
		github.com/uber/tchannel-go.registerNewChannel(0xc420279520)
			/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x6e
		github.com/uber/tchannel-go.NewChannel(0xb32f09, 0x5, 0xc4200c6420, 0xb3261b, 0x1, 0xc421155f9a)
			/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:292 +0xe78
		github.com/uber/tchannel-go/testutils.NewServerChannel(0xc4200c6000, 0xc4210ce690, 0xc4200c6000, 0x4159b0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:51 +0x615
		github.com/uber/tchannel-go/testutils.newServer(0xe34160, 0xc4210ce690, 0xc4200c6000, 0x10)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:71 +0x65
		github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc4210ce690, 0xb524f8, 0xc4200c6000, 0x0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:262 +0x72
		github.com/uber/tchannel-go/testutils.(*TestServer).addRelay(0xc4210ce690, 0xc4214d9940)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:243 +0x1e2
		github.com/uber/tchannel-go/testutils.NewTestServer(0xe342a0, 0xc420124690, 0xc4214d9940, 0x0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:92 +0x1f2
		github.com/uber/tchannel-go/testutils.withServer(0xe342a0, 0xc420124690, 0xc4214d9940, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:447 +0x51
		github.com/uber/tchannel-go/testutils.WithTestServer(0xe342a0, 0xc420124690, 0xc4205cccd8, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:127 +0x2b1
		github.com/uber/tchannel-go_test.TestRelayBasedSweep(0xc420124690)
			/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:202 +0x6d8
		testing.tRunner(0xc420124690, 0xb52a70)
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16d
		created by testing.(*T).Run
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:789 +0x569
		}]
		20:44:36.401883 [I] Starting idle connections poller. [{serviceName relay} {process relay-40596} {chID 403} {idleCheckInterval 30s} {maxIdleTime 3m0s}]
		20:44:36.402244 [I] Channel is listening. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596}]
		20:44:36.404990 [I] Created new active connection. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.9} {remotePeerTChannelVersion 1.9.0-dev}]
		20:44:36.405172 [D] New active inbound connection for peer 127.0.0.1:42176 [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound}]
		20:44:36.407417 [I] Created new active connection. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.9} {remotePeerTChannelVersion 1.9.0-dev}]
		20:44:36.407563 [D] New active outbound connection for peer 127.0.0.1:43024 [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound}]
		20:44:36.408394 [D] Writing frame messageTypeCallReq[2] [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound}]
		20:44:36.410275 [D] Writing frame messageTypeCallRes[2] [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound}]
		20:44:46.413998 [D] Connection got EOF [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound}]
		20:44:46.414187 [I] Connection closing. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound} {reason network connection EOF}]
		20:44:46.414419 [D] Connection state updated in Close. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound} {newState connectionStartClose}]
		20:44:46.414647 [D] Removed peer from root peer list. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {remoteHostPort 127.0.0.1:42176}]
		20:44:46.415027 [D] Connection state updated during shutdown. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound} {newState connectionClosed}]
		20:44:46.415341 [D] stopping 0 exchanges due to error: tchannel error ErrCodeNetwork: EOF [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound} {exchange outbound}]
		20:44:46.415492 [D] stopping 0 exchanges due to error: tchannel error ErrCodeNetwork: EOF [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound} {exchange inbound}]
		20:44:46.415686 [D] Closing underlying network connection [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 495} {localAddr 127.0.0.1:40596} {remoteAddr 127.0.0.1:42176} {remoteHostPort 127.0.0.1:42176} {remoteIsEphemeral true} {remoteProcess testService-client-242} {connectionDirection inbound}]
		20:44:46.416726 [D] TEST: TestServer is closing and verifying channel [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596}]
		20:44:46.416786 [I] Channel.Close called. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596}]
		20:44:46.416918 [I] Stopping idle connections poller. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596}]
		20:44:46.416944 [D] Connection got EOF [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound}]
		20:44:46.417009 [I] Connection closing. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {reason channel closing}]
		20:44:46.417163 [I] Connection closing. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {reason network connection EOF}]
		20:44:46.417227 [D] Connection state updated in Close. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {newState connectionStartClose}]
		20:44:46.417356 [D] stopping 0 exchanges due to error: tchannel error ErrCodeNetwork: EOF [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {exchange outbound}]
		20:44:46.417499 [D] ConnectionCloseStateChange channel state = ChannelStartClose connection minState = connectionStartClose [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound}]
		20:44:46.417529 [D] stopping 0 exchanges due to error: tchannel error ErrCodeNetwork: EOF [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {exchange inbound}]
		20:44:46.417835 [D] Connection state updated during shutdown. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound} {newState connectionClosed}]
		20:44:46.417893 [D] Closing underlying network connection [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound}]
		20:44:46.418084 [D] ConnectionCloseStateChange channel state = ChannelClosed connection minState = connectionClosed [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596} {connID 498} {localAddr 127.0.0.1:34308} {remoteAddr 127.0.0.1:43024} {remoteHostPort 127.0.0.1:43024} {remoteIsEphemeral false} {remoteProcess testService-43024} {outboundHP 127.0.0.1:43024} {connectionDirection outbound}]
		20:44:46.418202 [I] Channel closed. [{serviceName relay} {process relay-40596} {chID 403} {hostPort 127.0.0.1:40596}]
	<autogenerated>:1: Debug logs:
		20:44:36.403246 [I] Created new channel. [{serviceName testService} {process testService-43024} {chID 404} {channelPtr 0xc420278000} {createdStack goroutine 3149 [running]:
		github.com/uber/tchannel-go.getStacks(0xc420278000, 0xb34f32, 0xb, 0xc4214d8c40)
			/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:435 +0xb7
		github.com/uber/tchannel-go.registerNewChannel(0xc420278000)
			/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x6e
		github.com/uber/tchannel-go.NewChannel(0xb34f32, 0xb, 0xc4200c69a0, 0xb3261b, 0x1, 0xc4204a809a)
			/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:292 +0xe78
		github.com/uber/tchannel-go/testutils.NewServerChannel(0xc4200c6840, 0xc4210ce690, 0xc4200c6840, 0x4159b0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:51 +0x615
		github.com/uber/tchannel-go/testutils.newServer(0xe34160, 0xc4210ce690, 0xc4200c6840, 0x148)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:71 +0x65
		github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc4210ce690, 0xb524f8, 0xc4200c6840, 0x46577c)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:262 +0x72
		github.com/uber/tchannel-go/testutils.(*TestServer).NewServer(0xc4210ce690, 0xc4214d9e20, 0xf)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:222 +0x104
		github.com/uber/tchannel-go_test.TestRelayBasedSweep.func1(0xc4210ce690)
			/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:207 +0x229
		github.com/uber/tchannel-go/testutils.withServer(0xe342a0, 0xc420124690, 0xc4214d9940, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:452 +0x93
		github.com/uber/tchannel-go/testutils.WithTestServer(0xe342a0, 0xc420124690, 0xc4205cccd8, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:127 +0x2b1
		github.com/uber/tchannel-go_test.TestRelayBasedSweep(0xc420124690)
			/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:202 +0x6d8
		testing.tRunner(0xc420124690, 0xb52a70)
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16d
		created by testing.(*T).Run
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:789 +0x569
		}]
		20:44:36.403365 [I] Channel is listening. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024}]
		20:44:36.407359 [I] Created new active connection. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.9} {remotePeerTChannelVersion 1.9.0-dev}]
		20:44:36.407550 [D] New active inbound connection for peer 127.0.0.1:34308 [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
		20:44:36.408738 [D] Creating new inbound message exchange for [messageTypeCallReq:2] [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {exchange inbound}]
		20:44:36.409054 [D] Received incoming call for testService from 127.0.0.1:34308(relay-40596) [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {In-Call 2}]
		20:44:36.409333 [D] Dispatching testService:echo from 127.0.0.1:34308(relay-40596) [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
		20:44:36.409600 [D] Removing inbound message exchange 2 [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {exchange inbound}]
		20:44:36.409671 [D] Writing frame messageTypeCallRes[2] [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
		20:44:36.409860 [D] Wait for timeout/cancellation interrupted by error: mex has been shutdown [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {In-Call 2}]
		20:44:36.410039 [D] Removing inbound message exchange 2 due to timeout, cancellation or blackhole [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {exchange inbound}]
		20:44:46.413197 [D] TEST: TestServer is closing and verifying channel [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024}]
		20:44:46.413321 [I] Channel.Close called. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024}]
		20:44:46.413760 [I] Connection closing. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {reason channel closing}]
		20:44:46.413847 [D] Connection state updated in Close. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {newState connectionStartClose}]
		20:44:46.416004 [D] Removed peer from root peer list. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {remoteHostPort 127.0.0.1:34308}]
		20:44:46.416160 [D] ConnectionCloseStateChange channel state = ChannelStartClose connection minState = connectionStartClose [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
		20:44:46.416376 [D] Connection state updated during shutdown. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound} {newState connectionClosed}]
		20:44:46.416474 [D] Closing underlying network connection [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
		20:44:46.416526 [D] ConnectionCloseStateChange channel state = ChannelClosed connection minState = connectionClosed [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
		20:44:46.416660 [I] Channel closed. [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024}]
		20:44:46.416728 [D] Ignoring error after connection was closed: read tcp 127.0.0.1:43024->127.0.0.1:34308: use of closed network connection [{serviceName testService} {process testService-43024} {chID 404} {hostPort 127.0.0.1:43024} {connID 497} {localAddr 127.0.0.1:43024} {remoteAddr 127.0.0.1:34308} {remoteHostPort 127.0.0.1:34308} {remoteIsEphemeral true} {remoteProcess relay-40596} {connectionDirection inbound}]
	<autogenerated>:1: Debug logs:
		20:44:36.403958 [I] Created new channel. [{serviceName testService-client} {process testService-client-242} {chID 405} {channelPtr 0xc4202781a0} {createdStack goroutine 3149 [running]:
		github.com/uber/tchannel-go.getStacks(0xc420278100, 0xb3845d, 0x12, 0xc4214d8da0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:435 +0xb7
		github.com/uber/tchannel-go.registerNewChannel(0xc4202781a0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x6e
		github.com/uber/tchannel-go.NewChannel(0xb3845d, 0x12, 0xc4200c6f20, 0xb3261b, 0x1, 0xc4204a8154)
			/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:292 +0xe78
		github.com/uber/tchannel-go/testutils.NewClientChannel(0xc4200c6dc0, 0xc4210ce690, 0xc4200c6dc0, 0x4159b0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:74 +0x308
		github.com/uber/tchannel-go/testutils.newClient(0xe34160, 0xc4210ce690, 0xc4200c6dc0, 0x148)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:85 +0x65
		github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc4210ce690, 0xb524f0, 0xc4200c6dc0, 0x4)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:262 +0x72
		github.com/uber/tchannel-go/testutils.(*TestServer).NewClient(0xc4210ce690, 0xc4214d9e20, 0x0)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:217 +0xda
		github.com/uber/tchannel-go_test.TestRelayBasedSweep.func1(0xc4210ce690)
			/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:212 +0x272
		github.com/uber/tchannel-go/testutils.withServer(0xe342a0, 0xc420124690, 0xc4214d9940, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:452 +0x93
		github.com/uber/tchannel-go/testutils.WithTestServer(0xe342a0, 0xc420124690, 0xc4205cccd8, 0xc4214d9c98)
			/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:127 +0x2b1
		github.com/uber/tchannel-go_test.TestRelayBasedSweep(0xc420124690)
			/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:202 +0x6d8
		testing.tRunner(0xc420124690, 0xb52a70)
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16d
		created by testing.(*T).Run
			/home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:789 +0x569
		}]
		20:44:36.405189 [I] Created new active connection. [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.9} {remotePeerTChannelVersion 1.9.0-dev}]
		20:44:36.405332 [D] New active outbound connection for peer 127.0.0.1:40596 [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound}]
		20:44:36.405871 [D] Creating new outbound message exchange for [messageTypeCallReq:2] [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {exchange outbound}]
		20:44:36.406240 [D] Writing frame messageTypeCallReq[2] [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound}]
		20:44:36.410554 [D] forwarding outbound messageTypeCallRes[2] [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {exchange outbound}]
		20:44:36.410791 [D] Removing outbound message exchange 2 [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {exchange outbound}]
		20:44:46.411895 [D] TEST: TestServer is closing and verifying channel [{serviceName testService-client} {process testService-client-242} {chID 405}]
		20:44:46.411930 [I] Channel.Close called. [{serviceName testService-client} {process testService-client-242} {chID 405}]
		20:44:46.411979 [I] Connection closing. [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {reason channel closing}]
		20:44:46.412071 [D] Connection state updated in Close. [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {newState connectionStartClose}]
		20:44:46.412276 [D] Removed peer from root peer list. [{serviceName testService-client} {process testService-client-242} {chID 405} {remoteHostPort 127.0.0.1:40596}]
		20:44:46.412810 [D] ConnectionCloseStateChange channel state = ChannelStartClose connection minState = connectionStartClose [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound}]
		20:44:46.413008 [D] Connection state updated during shutdown. [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound} {newState connectionClosed}]
		20:44:46.413089 [D] ConnectionCloseStateChange channel state = ChannelClosed connection minState = connectionClosed [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound}]
		20:44:46.413165 [I] Channel closed. [{serviceName testService-client} {process testService-client-242} {chID 405}]
		20:44:46.413442 [D] Closing underlying network connection [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound}]
		20:44:46.413657 [D] Ignoring error after connection was closed: read tcp 127.0.0.1:42176->127.0.0.1:40596: use of closed network connection [{serviceName testService-client} {process testService-client-242} {chID 405} {connID 496} {localAddr 127.0.0.1:42176} {remoteAddr 127.0.0.1:40596} {remoteHostPort 127.0.0.1:40596} {remoteIsEphemeral false} {remoteProcess relay-40596} {outboundHP 127.0.0.1:40596} {connectionDirection outbound}]
```

After [increasing](https://github.com/uber/tchannel-go/pull/738) the timeout, the error has changed,
```
--- FAIL: TestServerBasedSweep (20.02s)
    --- FAIL: TestServerBasedSweep/no_relay (20.02s)
        logger.go:192: Unexpected log: [Error]: Skip closing idle Connection as it has pending calls. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
        logger.go:192: Unexpected log: [Error]: Skip closing idle Connection as it has pending calls. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
        logger.go:192: Unexpected log: [Error]: Skip closing idle Connection as it has pending calls. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
        idle_sweep_test.go:63: Some connections are still open: testService-46427: 1 open, testService-client-251: 1 open
        logger.go:165: Debug logs:
            23:08:16.358800 [I] Created new channel. [{serviceName testService} {process testService-46427} {chID 425} {channelPtr 0xc0003baa80} {createdStack goroutine 2596 [running]:
            github.com/uber/tchannel-go.getStacks(0xc0003baa00, 0xbe6081, 0xb, 0xbe7a2b)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:439 +0xd5
            github.com/uber/tchannel-go.registerNewChannel(0xc0003baa80)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x86
            github.com/uber/tchannel-go.NewChannel(0xbe6081, 0xb, 0xc0004af080, 0xbe34ab, 0x1, 0xc0001d904a)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:294 +0xfa1
            github.com/uber/tchannel-go/testutils.NewServerChannel(0xc0004aef20, 0xc0011b6540, 0xc0004aef20, 0x203000)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:52 +0x67d
            github.com/uber/tchannel-go/testutils.newServer(0xcd13a0, 0xc0011b6540, 0xc0004aef20, 0x0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:71 +0x65
            github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc0011b6540, 0xc06b00, 0xc0004aef20, 0x4274a0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:295 +0x72
            github.com/uber/tchannel-go/testutils.(*TestServer).NewServer(0xc0011b6540, 0xc000067c98, 0x0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:255 +0x100
            github.com/uber/tchannel-go/testutils.NewTestServer(0xcd14e0, 0xc000242200, 0xc0000a7c98, 0xc00002a218)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:87 +0x196
            github.com/uber/tchannel-go/testutils.withServer(0xcd14e0, 0xc000242200, 0xc000050c98, 0xc00137a580)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:465 +0x51
            github.com/uber/tchannel-go/testutils.WithTestServer.func1(0xcd14e0, 0xc000242200)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:131 +0x112
            github.com/uber/tchannel-go/testutils.runSubTest.func1(0xc000242200)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:101 +0x53
            testing.tRunner(0xc000242200, 0xc00040a760)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:865 +0x164
            created by testing.(*T).Run
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:916 +0x65b
            }]
            23:08:16.358928 [I] Starting idle connections poller. [{serviceName testService} {process testService-46427} {chID 425} {idleCheckInterval 30s} {maxIdleTime 3m0s}]
            23:08:16.359290 [I] Channel is listening. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427}]
            23:08:16.362785 [I] Created new active connection. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.12.4} {remotePeerTChannelVersion 1.13.0-dev}]
            23:08:16.371282 [D] New active inbound connection for peer 127.0.0.1:45068 [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
            23:08:16.375322 [D] Creating new inbound message exchange for [messageTypeCallReq:2] [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound} {exchange inbound}]
            23:08:16.376031 [D] Received incoming call for testService from 127.0.0.1:45068(testService-client-251) [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound} {In-Call 2}]
            23:08:16.376779 [D] Dispatching testService:echo from 127.0.0.1:45068(testService-client-251) [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
            23:08:16.377639 [D] Writing frame messageTypeCallRes[2] [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
            23:08:16.379584 [E] Skip closing idle Connection as it has pending calls. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
            23:08:16.379903 [E] Skip closing idle Connection as it has pending calls. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
            23:08:16.380153 [E] Skip closing idle Connection as it has pending calls. [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound}]
            23:08:16.380423 [D] Removing inbound message exchange 2 [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound} {exchange inbound}]
            23:08:16.380461 [D] Wait for timeout/cancellation interrupted by error: mex has been shutdown [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound} {In-Call 2}]
            23:08:16.380656 [D] Removing inbound message exchange 2 due to timeout, cancellation or blackhole [{serviceName testService} {process testService-46427} {chID 425} {hostPort 127.0.0.1:46427} {connID 429} {localAddr 127.0.0.1:46427} {remoteAddr 127.0.0.1:45068} {remoteHostPort 127.0.0.1:45068} {remoteIsEphemeral true} {remoteProcess testService-client-251} {connectionDirection inbound} {exchange inbound}]
        logger.go:165: Debug logs:
            23:08:16.360137 [I] Created new channel. [{serviceName testService-client} {process testService-client-251} {chID 426} {channelPtr 0xc0003bae00} {createdStack goroutine 2596 [running]:
            github.com/uber/tchannel-go.getStacks(0xc0003bae00, 0xbe99c2, 0x12, 0xbe7a2b)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:439 +0xd5
            github.com/uber/tchannel-go.registerNewChannel(0xc0003bae00)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x86
            github.com/uber/tchannel-go.NewChannel(0xbe99c2, 0x12, 0xc0004af4a0, 0xbe34ab, 0x1, 0xc0001d90e4)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:294 +0xfa1
            github.com/uber/tchannel-go/testutils.NewClientChannel(0xc0004af1e0, 0xc0011b6540, 0xc0004af1e0, 0x203000)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:75 +0x2d4
            github.com/uber/tchannel-go/testutils.newClient(0xcd13a0, 0xc0011b6540, 0xc0004af1e0, 0xc00060cb20)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:85 +0x65
            github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc0011b6540, 0xc06af8, 0xc0004af1e0, 0x4)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:295 +0x72
            github.com/uber/tchannel-go/testutils.(*TestServer).NewClient(0xc0011b6540, 0xc0004ae9a0, 0x0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:250 +0xd0
            github.com/uber/tchannel-go_test.TestServerBasedSweep.func1(0xcd14e0, 0xc000242200, 0xc0011b6540)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:123 +0x12a
            github.com/uber/tchannel-go/testutils.withServer(0xcd14e0, 0xc000242200, 0xc000050c98, 0xc00137a580)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:470 +0xa7
            github.com/uber/tchannel-go/testutils.WithTestServer.func1(0xcd14e0, 0xc000242200)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:131 +0x112
            github.com/uber/tchannel-go/testutils.runSubTest.func1(0xc000242200)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:101 +0x53
            testing.tRunner(0xc000242200, 0xc00040a760)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:865 +0x164
            created by testing.(*T).Run
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:916 +0x65b
            }]
            23:08:16.371644 [I] Created new active connection. [{serviceName testService-client} {process testService-client-251} {chID 426} {connID 430} {localAddr 127.0.0.1:45068} {remoteAddr 127.0.0.1:46427} {remoteHostPort 127.0.0.1:46427} {remoteIsEphemeral false} {remoteProcess testService-46427} {outboundHP 127.0.0.1:46427} {connectionDirection outbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.12.4} {remotePeerTChannelVersion 1.13.0-dev}]
            23:08:16.371907 [D] New active outbound connection for peer 127.0.0.1:46427 [{serviceName testService-client} {process testService-client-251} {chID 426} {connID 430} {localAddr 127.0.0.1:45068} {remoteAddr 127.0.0.1:46427} {remoteHostPort 127.0.0.1:46427} {remoteIsEphemeral false} {remoteProcess testService-46427} {outboundHP 127.0.0.1:46427} {connectionDirection outbound}]
            23:08:16.373567 [D] Creating new outbound message exchange for [messageTypeCallReq:2] [{serviceName testService-client} {process testService-client-251} {chID 426} {connID 430} {localAddr 127.0.0.1:45068} {remoteAddr 127.0.0.1:46427} {remoteHostPort 127.0.0.1:46427} {remoteIsEphemeral false} {remoteProcess testService-46427} {outboundHP 127.0.0.1:46427} {connectionDirection outbound} {exchange outbound}]
            23:08:16.374712 [D] Writing frame messageTypeCallReq[2] [{serviceName testService-client} {process testService-client-251} {chID 426} {connID 430} {localAddr 127.0.0.1:45068} {remoteAddr 127.0.0.1:46427} {remoteHostPort 127.0.0.1:46427} {remoteIsEphemeral false} {remoteProcess testService-46427} {outboundHP 127.0.0.1:46427} {connectionDirection outbound}]
            23:08:16.378095 [D] forwarding outbound messageTypeCallRes[2] [{serviceName testService-client} {process testService-client-251} {chID 426} {connID 430} {localAddr 127.0.0.1:45068} {remoteAddr 127.0.0.1:46427} {remoteHostPort 127.0.0.1:46427} {remoteIsEphemeral false} {remoteProcess testService-46427} {outboundHP 127.0.0.1:46427} {connectionDirection outbound} {exchange outbound}]
            23:08:16.378857 [D] Removing outbound message exchange 2 [{serviceName testService-client} {process testService-client-251} {chID 426} {connID 430} {localAddr 127.0.0.1:45068} {remoteAddr 127.0.0.1:46427} {remoteHostPort 127.0.0.1:46427} {remoteIsEphemeral false} {remoteProcess testService-46427} {outboundHP 127.0.0.1:46427} {connectionDirection outbound} {exchange outbound}]
=== RUN   TestClientBasedSweep
=== RUN   TestClientBasedSweep/no_relay
--- FAIL: TestClientBasedSweep (1.06s)
    --- FAIL: TestClientBasedSweep/no_relay (1.06s)
        test_server.go:409: found unexpected goroutines:
            [Goroutine 2597 in state select, with github.com/uber/tchannel-go.(*idleSweep).pollerLoop on top of the stack:
            goroutine 2597 [select]:
            github.com/uber/tchannel-go.(*idleSweep).pollerLoop(0xc000f6dc50)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep.go:81 +0x1e7
            created by github.com/uber/tchannel-go.(*idleSweep).start
            	/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep.go:63 +0x2a9
            
             Goroutine 2598 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
            goroutine 2598 [IO wait]:
            internal/poll.runtime_pollWait(0x2b99ed37c240, 0x72, 0x72)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/runtime/netpoll.go:182 +0x56
            internal/poll.(*pollDesc).wait(0xc00044cc98, 0x72, 0x0, 0x0, 0xbe4504)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:87 +0xe5
            internal/poll.(*pollDesc).waitRead(...)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:92
            internal/poll.(*FD).Accept(0xc00044cc80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_unix.go:384 +0x294
            net.(*netFD).accept(0xc00044cc80, 0x2b99f8050c88, 0xc0001d9098, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/fd_unix.go:238 +0x56
            net.(*TCPListener).accept(0xc000156118, 0x4a58c5, 0xc0001d9098, 0xffffffff)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/tcpsock_posix.go:139 +0x50
            net.(*TCPListener).Accept(0xc000156118, 0xc06b28, 0xc000240e80, 0xc0003b3980, 0x91e6c8)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/tcpsock.go:260 +0x51
            github.com/uber/tchannel-go/tnet.(*listener).Accept(0xc000240e80, 0x0, 0x0, 0x0, 0x0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/tnet/listener.go:80 +0x96
            github.com/uber/tchannel-go.(*Channel).serve(0xc0003baa80)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:460 +0xb1
            created by github.com/uber/tchannel-go.(*Channel).Serve
            	/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:333 +0x571
            
             Goroutine 2602 in state select, with github.com/uber/tchannel-go.(*Connection).writeFrames on top of the stack:
            goroutine 2602 [select]:
            github.com/uber/tchannel-go.(*Connection).writeFrames(0xc0003f6000, 0xc0000001ad)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:703 +0x185
            created by github.com/uber/tchannel-go.(*Channel).newConnection
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:342 +0x13c7
            
             Goroutine 2619 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
            goroutine 2619 [IO wait]:
            internal/poll.runtime_pollWait(0x2b99ed330b08, 0x72, 0xcbfc00)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/runtime/netpoll.go:182 +0x56
            internal/poll.(*pollDesc).wait(0xc00014a398, 0x72, 0x0, 0x10, 0xffffffffffffffff)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:87 +0xe5
            internal/poll.(*pollDesc).waitRead(...)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:92
            internal/poll.(*FD).Read(0xc00014a380, 0xc0012ac000, 0x10, 0x10, 0x0, 0x0, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_unix.go:169 +0x221
            net.(*netFD).Read(0xc00014a380, 0xc0012ac000, 0x10, 0x10, 0x0, 0xc000210180, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/fd_unix.go:202 +0x66
            net.(*conn).Read(0xc000156158, 0xc0012ac000, 0x10, 0x10, 0x10d7350, 0xc000000001, 0x6e)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/net.go:177 +0xa2
            io.ReadAtLeast(0xcbeb40, 0xc000156158, 0xc0012ac000, 0x10, 0x10, 0x10, 0x0, 0x0, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/io/io.go:310 +0x96
            io.ReadFull(...)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/io/io.go:329
            github.com/uber/tchannel-go.(*Connection).readFrames(0xc0000c4f00, 0xc0000001ae)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:627 +0x174
            created by github.com/uber/tchannel-go.(*Channel).newConnection
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:341 +0x1397
            
             Goroutine 2620 in state select, with github.com/uber/tchannel-go.(*Connection).writeFrames on top of the stack:
            goroutine 2620 [select]:
            github.com/uber/tchannel-go.(*Connection).writeFrames(0xc0000c4f00, 0xc0000001ae)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:703 +0x185
            created by github.com/uber/tchannel-go.(*Channel).newConnection
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:342 +0x13c7
            
             Goroutine 2601 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
            goroutine 2601 [IO wait]:
            internal/poll.runtime_pollWait(0x2b99ed330558, 0x72, 0xcbfc00)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/runtime/netpoll.go:182 +0x56
            internal/poll.(*pollDesc).wait(0xc00014a418, 0x72, 0x0, 0x10, 0xffffffffffffffff)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:87 +0xe5
            internal/poll.(*pollDesc).waitRead(...)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:92
            internal/poll.(*FD).Read(0xc00014a400, 0xc0001d8000, 0x10, 0x10, 0x0, 0x0, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/internal/poll/fd_unix.go:169 +0x221
            net.(*netFD).Read(0xc00014a400, 0xc0001d8000, 0x10, 0x10, 0xc0000e6050, 0xc0000e6010, 0xc0000e6088)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/fd_unix.go:202 +0x66
            net.(*conn).Read(0xc000156150, 0xc0001d8000, 0x10, 0x10, 0x110af20, 0xc0003f6040, 0xc000240a20)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/net/net.go:177 +0xa2
            io.ReadAtLeast(0xcbeb40, 0xc000156150, 0xc0001d8000, 0x10, 0x10, 0x10, 0x0, 0x0, 0x0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/io/io.go:310 +0x96
            io.ReadFull(...)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/io/io.go:329
            github.com/uber/tchannel-go.(*Connection).readFrames(0xc0003f6000, 0xc0000001ad)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:627 +0x174
            created by github.com/uber/tchannel-go.(*Channel).newConnection
            	/home/travis/gopath/src/github.com/uber/tchannel-go/connection.go:341 +0x1397
            ]
        logger.go:165: Debug logs:
            23:08:36.383282 [I] Created new channel. [{serviceName testService} {process testService-33964} {chID 427} {channelPtr 0xc0001a6700} {createdStack goroutine 2623 [running]:
            github.com/uber/tchannel-go.getStacks(0xc0001a6700, 0xbe6081, 0xb, 0xbe7a2b)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:439 +0xd5
            github.com/uber/tchannel-go.registerNewChannel(0xc0001a6700)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x86
            github.com/uber/tchannel-go.NewChannel(0xbe6081, 0xb, 0xc0000de580, 0xbe34ab, 0x1, 0xc0012ac2ca)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:294 +0xfa1
            github.com/uber/tchannel-go/testutils.NewServerChannel(0xc0000de420, 0xc000f60310, 0xc0000de420, 0x203000)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:52 +0x67d
            github.com/uber/tchannel-go/testutils.newServer(0xcd13a0, 0xc000f60310, 0xc0000de420, 0xb36980)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:71 +0x65
            github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc000f60310, 0xc06b00, 0xc0000de420, 0x4274a0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:295 +0x72
            github.com/uber/tchannel-go/testutils.(*TestServer).NewServer(0xc000f60310, 0xc000065c98, 0xc000050be8)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:255 +0x100
            github.com/uber/tchannel-go/testutils.NewTestServer(0xcd14e0, 0xc0000faa00, 0xc0013e9c98, 0xc00002a218)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:87 +0x196
            github.com/uber/tchannel-go/testutils.withServer(0xcd14e0, 0xc0000faa00, 0xc000050c98, 0xc00012c440)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:465 +0x51
            github.com/uber/tchannel-go/testutils.WithTestServer.func1(0xcd14e0, 0xc0000faa00)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:131 +0x112
            github.com/uber/tchannel-go/testutils.runSubTest.func1(0xc0000faa00)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:101 +0x53
            testing.tRunner(0xc0000faa00, 0xc0004292a0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:865 +0x164
            created by testing.(*T).Run
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:916 +0x65b
            }]
            23:08:36.383433 [I] Channel is listening. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964}]
            23:08:36.387810 [I] Created new active connection. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.12.4} {remotePeerTChannelVersion 1.13.0-dev}]
            23:08:36.388118 [D] New active inbound connection for peer 127.0.0.1:59100 [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
            23:08:36.393142 [D] Creating new inbound message exchange for [messageTypeCallReq:2] [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {exchange inbound}]
            23:08:36.394133 [D] Received incoming call for testService from 127.0.0.1:59100(testService-client-252) [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {In-Call 2}]
            23:08:36.396794 [D] Dispatching testService:echo from 127.0.0.1:59100(testService-client-252) [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
            23:08:36.398044 [D] Removing inbound message exchange 2 [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {exchange inbound}]
            23:08:36.398560 [D] Wait for timeout/cancellation interrupted by error: mex has been shutdown [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {In-Call 2}]
            23:08:36.398841 [D] Removing inbound message exchange 2 due to timeout, cancellation or blackhole [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {exchange inbound}]
            23:08:36.399221 [D] Writing frame messageTypeCallRes[2] [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
            23:08:36.410309 [D] Connection got EOF [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
            23:08:36.410486 [I] Connection closing. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {reason network connection EOF}]
            23:08:36.410635 [D] Connection state updated in Close. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {newState connectionStartClose}]
            23:08:36.410903 [D] Removed peer from root peer list. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {remoteHostPort 127.0.0.1:59100}]
            23:08:36.412942 [D] TEST: Test function complete [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964}]
            23:08:36.413315 [D] TEST: TestServer is verifying channel [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964}]
            23:08:36.414097 [D] TEST: TestServer is closing and verifying channel [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964}]
            23:08:36.414149 [I] Channel.Close called. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964}]
            23:08:36.414381 [I] Connection closing. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {reason channel closing}]
            23:08:36.417921 [D] ConnectionCloseStateChange channel state = ChannelStartClose connection minState = connectionStartClose [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
            23:08:36.418647 [D] Connection state updated during shutdown. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {newState connectionClosed}]
            23:08:36.423656 [D] ConnectionCloseStateChange channel state = ChannelClosed connection minState = connectionClosed [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
            23:08:36.423850 [I] Channel closed. [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964}]
            23:08:36.423983 [D] stopping 0 exchanges due to error: tchannel error ErrCodeNetwork: EOF [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {exchange outbound}]
            23:08:36.424149 [D] stopping 0 exchanges due to error: tchannel error ErrCodeNetwork: EOF [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound} {exchange inbound}]
            23:08:36.426462 [D] Closing underlying network connection [{serviceName testService} {process testService-33964} {chID 427} {hostPort 127.0.0.1:33964} {connID 431} {localAddr 127.0.0.1:33964} {remoteAddr 127.0.0.1:59100} {remoteHostPort 127.0.0.1:59100} {remoteIsEphemeral true} {remoteProcess testService-client-252} {connectionDirection inbound}]
        logger.go:165: Debug logs:
            23:08:36.384654 [I] Created new channel. [{serviceName testService-client} {process testService-client-252} {chID 428} {channelPtr 0xc0001a68c0} {createdStack goroutine 2623 [running]:
            github.com/uber/tchannel-go.getStacks(0xc0001a6800, 0xbe99c2, 0x12, 0xbe7a2b)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/introspection.go:439 +0xd5
            github.com/uber/tchannel-go.registerNewChannel(0xc0001a68c0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/all_channels.go:39 +0x86
            github.com/uber/tchannel-go.NewChannel(0xbe99c2, 0x12, 0xc0000de840, 0xbe34ab, 0x1, 0xc0012ac354)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/channel.go:294 +0xfa1
            github.com/uber/tchannel-go/testutils.NewClientChannel(0xc0000de6e0, 0xc000f60310, 0xc0000de6e0, 0x203000)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel.go:75 +0x2d4
            github.com/uber/tchannel-go/testutils.newClient(0xcd13a0, 0xc000f60310, 0xc0000de6e0, 0xc0001346c0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/channel_t.go:85 +0x65
            github.com/uber/tchannel-go/testutils.(*TestServer).addChannel(0xc000f60310, 0xc06af8, 0xc0000de6e0, 0x4)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:295 +0x72
            github.com/uber/tchannel-go/testutils.(*TestServer).NewClient(0xc000f60310, 0xc0000de160, 0x0)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:250 +0xd0
            github.com/uber/tchannel-go_test.TestClientBasedSweep.func1(0xcd14e0, 0xc0000faa00, 0xc000f60310)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/idle_sweep_test.go:166 +0x12a
            github.com/uber/tchannel-go/testutils.withServer(0xcd14e0, 0xc0000faa00, 0xc000050c98, 0xc00012c440)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:470 +0xa7
            github.com/uber/tchannel-go/testutils.WithTestServer.func1(0xcd14e0, 0xc0000faa00)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:131 +0x112
            github.com/uber/tchannel-go/testutils.runSubTest.func1(0xc0000faa00)
            	/home/travis/gopath/src/github.com/uber/tchannel-go/testutils/test_server.go:101 +0x53
            testing.tRunner(0xc0000faa00, 0xc0004292a0)
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:865 +0x164
            created by testing.(*T).Run
            	/home/travis/.gimme/versions/go1.12.4.linux.amd64/src/testing/testing.go:916 +0x65b
            }]
            23:08:36.384796 [I] Starting idle connections poller. [{serviceName testService-client} {process testService-client-252} {chID 428} {idleCheckInterval 30s} {maxIdleTime 3m0s}]
            23:08:36.389642 [I] Created new active connection. [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {remotePeerLanguage go} {remotePeerLanguageVersion 1.12.4} {remotePeerTChannelVersion 1.13.0-dev}]
            23:08:36.389910 [D] New active outbound connection for peer 127.0.0.1:33964 [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound}]
            23:08:36.391061 [D] Creating new outbound message exchange for [messageTypeCallReq:2] [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {exchange outbound}]
            23:08:36.392461 [D] Writing frame messageTypeCallReq[2] [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound}]
            23:08:36.399752 [D] forwarding outbound messageTypeCallRes[2] [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {exchange outbound}]
            23:08:36.400943 [D] Removing outbound message exchange 2 [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {exchange outbound}]
            23:08:36.403896 [I] Closing idle inbound connection. [{serviceName testService-client} {process testService-client-252} {chID 428} {remotePeer 127.0.0.1:33964(testService-33964)} {lastActivityTime 2019-05-02 23:08:36.382221973 +0000 UTC}]
            23:08:36.407826 [I] Connection closing. [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {reason Idle connection closed}]
            23:08:36.407993 [D] Connection state updated in Close. [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {newState connectionStartClose}]
            23:08:36.408284 [D] Removed peer from root peer list. [{serviceName testService-client} {process testService-client-252} {chID 428} {remoteHostPort 127.0.0.1:33964}]
            23:08:36.408668 [D] Connection state updated during shutdown. [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound} {newState connectionClosed}]
            23:08:36.408682 [D] Closing underlying network connection [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound}]
            23:08:36.408946 [D] Ignoring error after connection was closed: read tcp 127.0.0.1:59100->127.0.0.1:33964: use of closed network connection [{serviceName testService-client} {process testService-client-252} {chID 428} {connID 432} {localAddr 127.0.0.1:59100} {remoteAddr 127.0.0.1:33964} {remoteHostPort 127.0.0.1:33964} {remoteIsEphemeral false} {remoteProcess testService-33964} {outboundHP 127.0.0.1:33964} {connectionDirection outbound}]
            23:08:36.413035 [D] TEST: TestServer is verifying channel [{serviceName testService-client} {process testService-client-252} {chID 428}]
            23:08:36.413633 [D] TEST: TestServer is closing and verifying channel [{serviceName testService-client} {process testService-client-252} {chID 428}]
            23:08:36.413688 [I] Channel.Close called. [{serviceName testService-client} {process testService-client-252} {chID 428}]
            23:08:36.413750 [I] Stopping idle connections poller. [{serviceName testService-client} {process testService-client-252} {chID 428}]
            23:08:36.413884 [I] Channel closed. [{serviceName testService-client} {process testService-client-252} {chID 428}]
```

--- a/connection_test.go
+++ b/connection_test.go
@@ -920,7 +920,7 @@ func TestConnectionIDs(t *testing.T) {
 
 		// We want to reuse the same connection for the rest of the test which
 		// only makes sense when the relay is not used.
-		if ts.Relay() != nil {
+		if ts.HasRelay() {
 			return
 		}
 

--- a/idle_sweep_test.go
+++ b/idle_sweep_test.go
@@ -184,9 +184,6 @@ func TestClientBasedSweep(t *testing.T) {
 // outbound.
 func TestRelayBasedSweep(t *testing.T) {
 	listener := newPeerStatusListener()
-	ctx, cancel := NewContext(time.Second)
-	defer cancel()
-
 	relayTicker := testutils.NewFakeTicker()
 	clock := testutils.NewStubClock(time.Now())
 
@@ -199,20 +196,16 @@ func TestRelayBasedSweep(t *testing.T) {
 		SetMaxIdleTime(3 * time.Minute).
 		SetIdleCheckInterval(30 * time.Second).
 		SetOnPeerStatusChanged(listener.onStatusChange).
+		SetDisableServer(). // We create our own server without the idle sweeper.
 		SetRelayOnly()
-
 	testutils.WithTestServer(t, relayOpts, func(t testing.TB, ts *testutils.TestServer) {
-		// Replace the auto-created server with a new one that doesn't have
-		// an idle-connection poller.
-		ts.Relay().GetSubChannel(ts.ServiceName()).Peers().Remove(
-			ts.Server().PeerInfo().HostPort)
 		server := ts.NewServer(opts)
 		testutils.RegisterEcho(server, nil)
 
 		// Make a call to the server via relay, which will establish connections:
 		// Client -> Relay -> Server
 		client := ts.NewClient(opts)
-		raw.Call(ctx, client, ts.HostPort(), ts.ServiceName(), "echo", nil, nil)
+		testutils.AssertEcho(t, client, ts.HostPort(), server.ServiceName())
 
 		relayTicker.Tick()
 
@@ -341,7 +334,8 @@ func TestIdleSweepIgnoresConnectionsWithCalls(t *testing.T) {
 			},
 		}
 
-		if relay := ts.Relay(); relay != nil {
+		if ts.HasRelay() {
+			relay := ts.Relay()
 			check.ch = relay
 			check.preCloseConns++
 			oldTick := check.tick

--- a/testutils/channel_opts.go
+++ b/testutils/channel_opts.go
@@ -55,6 +55,10 @@ type ChannelOpts struct {
 	// By default, all tests are run with a relay interposed.
 	DisableRelay bool
 
+	// DisableServer disables creation of the TChannel server.
+	// This is typically only used in relay tests when a custom server is required.
+	DisableServer bool
+
 	// OnlyRelay instructs TestServer the test must only be run with a relay.
 	OnlyRelay bool
 
@@ -172,6 +176,13 @@ func (o *ChannelOpts) NoRelay() *ChannelOpts {
 // SetRelayOnly instructs TestServer to only run with a relay in front of this channel.
 func (o *ChannelOpts) SetRelayOnly() *ChannelOpts {
 	o.OnlyRelay = true
+	return o
+}
+
+// SetDisableServer disables creation of the TChannel server.
+// This is typically only used in relay tests when a custom server is required.
+func (o *ChannelOpts) SetDisableServer() *ChannelOpts {
+	o.DisableServer = true
 	return o
 }
 

--- a/testutils/test_server.go
+++ b/testutils/test_server.go
@@ -33,6 +33,7 @@ import (
 	"go.uber.org/multierr"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"go.uber.org/atomic"
 	"golang.org/x/net/context"
 )
@@ -45,8 +46,9 @@ var _leakedGoroutine atomic.Bool
 type TestServer struct {
 	testing.TB
 
-	// relayIdx is the index of the relay channel, if any, in the channels slice.
-	relayIdx int
+	// References to specific channels (if any, as they can be disabled)
+	relayCh  *tchannel.Channel
+	serverCh *tchannel.Channel
 
 	// relayHost is the relayer's StubRelayHost (if any).
 	relayHost *relaytest.StubRelayHost
@@ -84,10 +86,12 @@ func NewTestServer(t testing.TB, opts *ChannelOpts) *TestServer {
 		},
 	}
 
-	// Remove any relay options, since those should only be applied to addRelay.
-	serverOpts := opts.Copy()
-	serverOpts.RelayHost = nil
-	ts.NewServer(serverOpts)
+	if !opts.DisableServer {
+		// Remove any relay options, since those should only be applied to addRelay.
+		serverOpts := opts.Copy()
+		serverOpts.RelayHost = nil
+		ts.serverCh = ts.NewServer(serverOpts)
+	}
 
 	if opts == nil || !opts.DisableRelay {
 		ts.addRelay(opts)
@@ -160,6 +164,12 @@ func (ts *TestServer) SetVerifyOpts(opts *goroutines.VerifyOpts) {
 	ts.verifyOpts = opts
 }
 
+// HasServer returns whether this TestServer has a TChannel server, as
+// the server may have been disabled with the DisableServer option.
+func (ts *TestServer) HasServer() bool {
+	return ts.serverCh != nil
+}
+
 // Server returns the underlying TChannel for the server (i.e., the channel on
 // which we're registering handlers).
 //
@@ -167,15 +177,20 @@ func (ts *TestServer) SetVerifyOpts(opts *goroutines.VerifyOpts) {
 // callers should use the Client(), HostPort(), ServiceName(), and Register()
 // methods instead of accessing the server channel explicitly.
 func (ts *TestServer) Server() *tchannel.Channel {
-	return ts.channels[0]
+	require.True(ts, ts.HasServer(), "Cannot use Server as it was disabled")
+	return ts.serverCh
+}
+
+// HasRelay indicates whether this TestServer has a relay interposed between the
+// server and clients.
+func (ts *TestServer) HasRelay() bool {
+	return ts.relayCh != nil
 }
 
 // Relay returns the relay channel, if one is present.
 func (ts *TestServer) Relay() *tchannel.Channel {
-	if ts.HasRelay() {
-		return ts.channels[ts.relayIdx]
-	}
-	return nil
+	require.True(ts, ts.HasRelay(), "Cannot use Relay, not present in current test")
+	return ts.relayCh
 }
 
 // RelayHost returns the stub RelayHost for mapping service names to peers.
@@ -277,22 +292,14 @@ func (ts *TestServer) addRelay(parentOpts *ChannelOpts) {
 	opts.ServiceName = "relay"
 	opts.ChannelOptions.RelayHost = relayHost
 
-	ts.addChannel(newServer, opts)
-	if ts.relayHost != nil {
+	ts.relayCh = ts.addChannel(newServer, opts)
+	if ts.relayHost != nil && ts.HasServer() {
 		ts.relayHost.Add(ts.Server().ServiceName(), ts.Server().PeerInfo().HostPort)
 	}
 
 	if statter, ok := relayHost.(relayStatter); ok {
 		ts.relayStats = statter.Stats()
 	}
-
-	ts.relayIdx = len(ts.channels) - 1
-}
-
-// HasRelay indicates whether this TestServer has a relay interposed between the
-// server and clients.
-func (ts *TestServer) HasRelay() bool {
-	return ts.relayIdx > 0
 }
 
 func (ts *TestServer) addChannel(createChannel func(t testing.TB, opts *ChannelOpts) *tchannel.Channel, opts *ChannelOpts) *tchannel.Channel {
@@ -472,6 +479,8 @@ func withServer(t testing.TB, chanOpts *ChannelOpts, f func(testing.TB, *TestSer
 	defer ts.post()
 
 	f(t, ts)
-	ts.Server().Logger().Debugf("TEST: Test function complete")
+	if ts.HasServer() {
+		ts.Server().Logger().Debugf("TEST: Test function complete")
+	}
 	ts.CloseAndVerify()
 }

